### PR TITLE
add a richer type for `Message.application`

### PR DESCRIPTION
**Pull request by AbstractUmbra**
## Summary

This PR adds `MessageApplication` as a richer type for `Message.application`.

It also adds a new classmethod to `Asset` for getting app-icons cdn contents.
The `.icon` and `.cover_image` properties have not been tested currently, but have followed the API docs and payload types.

Resolves [this card](https://github.com/Rapptz/discord.py/projects/3#card-82881897).

## Checklist

- [ ] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/message.py [hunk 1]
```diff
@@ -1327,16 +1380,9 @@ class Message(PartialMessage, Hashable):
 
         - ``type``: An integer denoting the type of message activity being requested.
         - ``party_id``: The party ID associated with the party.
-    application: Optional[:class:`dict`]
+    application: Optional[:class:`~discord.MessageApplication`]
```
Comment (Rapptz) [inline discord/message.py:1383]: Needs a versionchanged

File: discord/message.py [hunk 2]
```diff
@@ -610,6 +612,57 @@ def flatten_handlers(cls: Type[Message]) -> Type[Message]:
...
+    Attributes
+    -----------
+    id: :class:`int`
+        The application ID.
+    description: :class:`str`
+        The application description.
+    name: :class:`str`
+        The application's name.
+
+    .. versionadded:: 2.0
```
Comment (Rapptz) [inline discord/message.py:627]: This belongs before Attributes

File: discord/message.py [hunk 3]
```diff
@@ -610,6 +612,57 @@ def flatten_handlers(cls: Type[Message]) -> Type[Message]:
...
+    """
+
+    __slots__ = (
+        '_state',
+        '_icon',
+        '_cover_image',
+        'id',
+        'description',
+        'name',
+    )
```
Comment (Rapptz) [inline discord/message.py:638]: I think these are small enough to be compacted into one line.

File: discord/message.py [hunk 4]
```diff
@@ -610,6 +612,57 @@ def flatten_handlers(cls: Type[Message]) -> Type[Message]:
...
+    def __init__(self, *, state: ConnectionState, data: MessageApplicationPayload) -> None:
+        self._state: ConnectionState = state
+        self.id: int = int(data['id'])
+        self.description: str = data['description']
+        self.name: str = data['name']
+        self._icon: Optional[str] = data['icon']
+        self._cover_image: Optional[str] = data.get('cover_image')
+
+    def __repr__(self) -> str:
+        return f"<MessageApplication id={self.id} name={self.name!r}>"
```
Comment (Rapptz) [inline discord/message.py:649]:
```suggestion
        return f'<MessageApplication id={self.id} name={self.name!r}>'
```

File: discord/message.py [hunk 5]
```diff
@@ -610,6 +612,57 @@ def flatten_handlers(cls: Type[Message]) -> Type[Message]:
...
+
+    @property
+    def icon(self) -> Optional[Asset]:
+        """Optional[:class:`Asset`]: The application's icon, if any."""
+        if self._icon:
+            return Asset._from_app_icon(state=self._state, object_id=self.id, icon_hash=self._icon, asset_type='icon')
```
Comment (Rapptz) [inline discord/message.py:655]: Please add an explicit None return.

File: discord/message.py [hunk 6]
```diff
@@ -610,6 +612,57 @@ def flatten_handlers(cls: Type[Message]) -> Type[Message]:
...
+
+    @property
+    def cover(self) -> Optional[Asset]:
+        """Optional[:class:`Asset`]: The application's cover image, if any."""
+        if self._cover_image:
+            return Asset._from_app_icon(
+                state=self._state, object_id=self.id, icon_hash=self._cover_image, asset_type='cover_image'
+            )
```
Comment (Rapptz) [inline discord/message.py:663]: Please add an explicit None return.

**Review by Rapptz — CHANGES_REQUESTED** [review]
Needs to be documented in the migrating.rst file

File: discord/asset.py [hunk 7]
```diff
@@ -254,6 +254,15 @@ def _from_icon(cls, state: _State, object_id: int, icon_hash: str, path: str) ->
             animated=False,
         )
 
+    @classmethod
+    def _from_app_icon(cls, state: _State, object_id: int, icon_hash: str, asset_type: Literal['icon', 'cover_image']):
```
Comment (NCPlayz) [inline discord/asset.py:258]:
```suggestion
    def _from_app_icon(cls, state: _State, object_id: int, icon_hash: str, asset_type: Literal['icon', 'cover_image']) -> Self
```

**Review by NCPlayz — CHANGES_REQUESTED** [review]

File: discord/message.py [hunk 8]
```diff
@@ -1327,16 +1374,12 @@ class Message(PartialMessage, Hashable):
 
         - ``type``: An integer denoting the type of message activity being requested.
         - ``party_id``: The party ID associated with the party.
-    application: Optional[:class:`dict`]
+    application: Optional[:class:`~discord.MessageApplication`]
         The rich presence enabled application associated with this message.
 
-        It is a dictionary with the following keys:
+        .. versionchanged:: 2.0
+            This now returns a richer interface type, not a `dict` of the raw payload.
```
Comment (Rapptz) [inline discord/message.py:1381]:
```suggestion
            Type is now :class:`MessageApplication` instead of :class:`dict`.
```

**Review by Rapptz — APPROVED** [review]

**Pull request merged**